A bunch of Lua/WFL API updates

CelticMinstrel opened the pull request:
I've been working on a bunch of mostly-small updates to the API. Some of these are adding little things that were missing, and I also put a lot of work into adding `__dir` metamethods in the Lua API to make more parts somewhat self-documenting. There'll probably still be more to come, but I thought I should run the CI before it gets _too_ long.

soliton- commented:
There is no explanation in what way changing how `wildcard_string_match` works works properly with all callers. Apart from making our version of globs/wildcards even more outlandish.

Reverting "Update and correct game config schema" wholesale makes no sense. If you want to combine actually combinable tag schemas then just do that.

ProditorMagnus commented on data/lua/core/stringx.lua:
```diff
@@ -61,6 +61,22 @@ function stringx.iter_ranges(str)
 	end)
 end
 
+---Test if a string begins with a specified prefix
```
is there already contains?

stevecotton commented on src/scripting/lua_attributes.hpp:
```diff
@@ -0,0 +1,162 @@
+	See the COPYING file for more details.
+*/
+
+/// New attribute registration system, mainly for objects with a lot of attributes, like units
```
"new" isn't a useful description in source code, for it will become "old that's still called new".

stevecotton commented on src/formula/callable_objects.hpp:
```diff
@@ -19,6 +19,7 @@
 #include "formula/formula.hpp"
 #include "map/location.hpp"
 #include "units/ptr.hpp"
+#include "color.hpp"
```
Alphabetical order please

stevecotton commented on src/formula/function.cpp:
```diff
@@ -24,6 +24,7 @@
 #include "pathutils.hpp"
 
 #include <boost/math/constants/constants.hpp>
+#include <boost/algorithm/string.hpp>
```
Alphabetical order please

gfgtdf commented on src/scripting/lua_attributes.hpp:
```diff
@@ -0,0 +1,162 @@
+#include <string_view>
+#include <vector>
+
+struct luaW_Registry {
```
Some comments would be nice, especially on what the Boolean parameters of the callback functions do

CelticMinstrel commented:
> There is no explanation in what way changing how `wildcard_string_match` works works properly with all callers. Apart from making our version of globs/wildcards even more outlandish.

Bash globs _do_ use commas as alternation though, albeit only in braced lists. Anyway, there are unit tests for `wildcard_string_match`, and they still pass, so…

> Reverting "Update and correct game config schema" wholesale makes no sense. If you want to combine actually combinable tag schemas then just do that.

I'm not sure how that would end up any different from just reverting it.

In any case, I'll probably split those two commits to a separate PR.

Vultraz commented on src/serialization/string_utils.cpp:
```diff
@@ -702,7 +702,12 @@ bool word_match(const std::string& message, const std::string& word) {
 }
 
 bool wildcard_string_match(const std::string& str, const std::string& match) {
```
This should take a string_view so substr doesn’t make new strings.

CelticMinstrel replied:
I'm resolving this for now because it doesn't belong in this PR. (The glob commit was supposed to be a separate PR – it was pushed with this by mistake.)

Vultraz commented on src/formula/function.cpp:
```diff
@@ -701,6 +725,39 @@ DEFINE_WFL_FUNCTION(lerp, 3, 3)
+	std::vector<variant> result;
+	result.reserve(colors.size());
+	for(auto clr : colors) {
+		result.push_back(variant(std::make_shared<color_callable>(clr)));
```
Emplace_back

soliton- commented:
> > There is no explanation in what way changing how `wildcard_string_match` works works properly with all callers. Apart from making our version of globs/wildcards even more outlandish.
> 
> Bash globs _do_ use commas as alternation though, albeit only in braced lists. Anyway, there are unit tests for `wildcard_string_match`, and they still pass, so…

You expect our unit tests represent all callers? Brace expansion in bash has nothing to do with globs at all. It's simple string generation.

Changing our definition of wildcards just so one caller doesn't need to split on comma is just unnecessary. It's already silly that we have an entirely useless extension of +.

> > Reverting "Update and correct game config schema" wholesale makes no sense. If you want to combine actually combinable tag schemas then just do that.
> 
> I'm not sure how that would end up any different from just reverting it.

I can only assume you didn't look at the commit.

stevecotton commented on data/lua/core/map.lua:
```diff
@@ -461,6 +461,9 @@ if wesnoth.kernel_type() == "Mapgen Lua Kernel" then
 	end, 'The filter is now the last parameter, instead of the radius')
+	wesnoth.map.vector_sum = wesnoth.deprecate_api('wesnoth.map.vector_sum', 'wesnoth.map.hex_vector_sum', 1, nil, wesnoth.map.hex_vector_sum)
+	wesnoth.map.vector_diff = wesnoth.deprecate_api('wesnoth.map.vector_diff', 'wesnoth.map.hex_vector_dif', 1, nil, wesnoth.map.hex_vector_diff)
+	wesnoth.map.vector_negation = wesnoth.deprecate_api('wesnoth.map.vector_negation', 'wesnoth.map.hex_vector_negation', 1, nil, wesnoth.map.hex_vector_negation)
```
Given that the old functions were already in `map`, it was obvious that they were something to do with the map. What's the advantage of renaming them with `hex_` ?

CelticMinstrel replied:
I thought the old names gave the impression that they represented simple vector arithmetic, like `(x1, y1) + (x2, y2) = (x1 + y1, x2 + y2)`.

stevecotton replied:
I assumed that these were using the same coordinates as the `hex_locations`, but it seems they're using x,y instead. So the new names with `hex_` in are worst than the old ones, even without the cost of deprecation.

CelticMinstrel replied:
I'm open to better suggestions, either for the functions or for `hex_location`. The key issue I have is that someone might see "vector" and think it's working in Cartesian space, but it's actually using hex-offset coordinates (the coordinates used by `map_location`). The name `hex_location` is _also_ perhaps not the best, as both `hex_location` and `map_location` are hex-based coordinates, just _two different_ types of hex-based coordinates. The `hex_location` version makes some algorithms easier, while the `map_location` version is better for addressing maps as a pseudo-rectangular grid.

stevecotton replied:
Leaving it as-is until a better name comes up is better than changing the name twice. I think the current name is the best option for now.

stevecotton commented on src/map/location.hpp:
```diff
@@ -27,6 +27,10 @@ class variable_set;
 
 struct wml_loc {};
 
+struct hex_location {
```
Needs documentation

stevecotton replied:
The name is bad, given that `hex` and `location` are both heavily used in this area of the code. `cubic_location` would be a bit better.

Still needs better documentation - or a link to good documentation. Saying that it's "also called cubic coordinates" isn't that helpful, because searching for that on the web returns a lot of pages talking about using three orthogonal axis for three-dimensional space, which then takes time to find the ones talking about using a 2d plane within that space.

Remember, even if it's the first result that you see when you do a web search, that can be because the search engine is using your browsing history.

CelticMinstrel replied:
Sure, I can add a link to redblobgames.com.

stevecotton commented:
> 53 files changed, 2734 insertions(+), 1022 deletions(-), counting lines that are changed in multiple commits only once

This is bound to end up with Github's "hide comments in the middle of discussion" feature making it hard to even work out what issues are still open. Please, make smaller PRs that can merge, instead of adding unrelated new features to a PR that's already been open for 3 days.

CelticMinstrel commented:
I rebased with whitespace fixes and removed the plugin stuff (it's kinda big and this was supposed to be a bunch of small changes). Assuming CI passes I intend to merge this as soon as possible (later today, or tomorrow at the latest), so this is the last call for reviews.

gfgtdf commented:
I think it would make sense to move even the smaller lua structures ,like current and scenario into their own file, not really a blocker tho

gfgtdf commented:
I'm also wondering whether it would be better to use a different container than std::map for the registry structures, like a flat_map for a hash map, since this is only about lookup speed

stevecotton commented on src/formula/function.cpp:
```diff
@@ -448,6 +449,29 @@ DEFINE_WFL_FUNCTION(replace, 3, 4)
+	std::string result = args()[0]->evaluate(variables, fdb).as_string();
+	std::string needle = args()[1]->evaluate(variables, fdb).as_string();
+	std::string replacement = args().back()->evaluate(variables, fdb).as_string();
+	boost::replace_all(result, needle, replacement);
```
The documentation for boost::replace_all doesn't answer my questions - does this correctly handle reallocating the string if replacement is longer than needle?

CelticMinstrel replied:
I can't tell either, but I would assume it would ultimately boil down to `string::replace` at some level, or `string::insert` at worst.

I did test it with a replacement longer than the needle, and it looks like it works.

stevecotton commented on src/formula/function.cpp:
```diff
@@ -701,6 +725,39 @@ DEFINE_WFL_FUNCTION(lerp, 3, 3)
+	const double alpha = args()[1]->evaluate(variables, fdb).as_decimal() / 1000.0;
+	// Same formula as red_to_green etc
+	const double val_scaled = std::clamp(0.01 * alpha, 0.0, 1.0);
+	const int idx = int(std::nearbyint((items.size() - 1) * val_scaled));
```
Missing a check that items.size() != zero.

CelticMinstrel replied:
I added the check to the Lua version of this function too.

stevecotton commented on src/scripting/lua_attributes.cpp:
```diff
@@ -0,0 +1,102 @@
+#include "scripting/lua_attributes.hpp"
+#include "scripting/lua_common.hpp"
+#include "scripting/push_check.hpp"
+#include "scripting/lua_kernel_base.hpp" // for luaW_get_attributes
```
Alphabetical order please

stevecotton commented on src/scripting/lua_attributes.cpp:
```diff
@@ -0,0 +1,102 @@
+/*
+	Copyright (C) 2009 - 2024
```
Is the 2009 correct here?

CelticMinstrel replied:
No.

stevecotton commented on src/scripting/lua_attributes.hpp:
```diff
@@ -0,0 +1,177 @@
+
+#include "config.hpp"
+#include "variable_info.hpp"
+#include "map/location.hpp"
```
Alphabetical order - I know it's annoying to be reminded, but please get in the habit of doing it

CelticMinstrel replied:
I do somewhat have a habit of doing it. The problem is that some of the files aren't already in alphabetical order, so I don't bother, and then I forget.

stevecotton commented on src/scripting/lua_attributes.hpp:
```diff
@@ -0,0 +1,177 @@
+
+/// Holds a lookup table for members of one type of object.
+struct luaW_Registry {
+	inline static std::map<std::string_view /* metatable */, std::reference_wrapper<luaW_Registry>> lookup;
```
What owns the strings that the string_views point to? If it's private_metatable or public_metatable, those are both destroyed before this member.

gfgtdf replied:
They are most likely status string literals.

gfgtdf replied:
Actually I might be wrong I didn't really read the code

CelticMinstrel replied:
They are static string literals. Those literals _are_ copied into `private_metatable` or `public_metatable`, but the string_view points to the original literal.

stevecotton commented on src/scripting/lua_unit.cpp:
```diff
@@ -597,6 +832,32 @@ static int impl_unit_status_set(lua_State *L)
 
+/**
+ * List statuses on a unit (__dir metamethod)
+ * This returns all known statuses (regardless of state) plus any currently set to true.
```
Does it avoid duplicates when a unit is affected by a known state?

CelticMinstrel replied:
This logic adds duplicates, yes. However, the `__dir` machinery in `wesnoth.print_attributes` automatically removes duplicates, so I thought it wasn't worth avoiding it here.

CelticMinstrel commented:
> I think it would make sense to move even the smaller lua structures ,like current and scenario into their own file, not really a blocker tho

I was actually thinking that too. I might do that later, as a refactor type of thing.
